HPX distributed map fails to compile

## Expected Behavior

I would expect that the distributed maps to behave in a similar manner to the partitioned vector. Specifically, in the way they are initialized.

## Actual Behavior

When attempting to register and initialize a distributed map, I receive the following compile error:

```
In file included from /Users/raac/Development/covid/mobility-cpp/src/connectivity.cpp:7:
In file included from /Users/raac/Development/vcpkg/installed/x64-osx/include/hpx/include/unordered_map.hpp:10:
/Users/raac/Development/vcpkg/installed/x64-osx/include/hpx/components/containers/unordered/unordered_map.hpp:485:47: error: no matching constructor for initialization of 'server::unordered_map_config_data'
            server::unordered_map_config_data data(partitions_);
                                              ^    ~~~~~~~~~~~
/Users/raac/Development/covid/mobility-cpp/src/connectivity.cpp:22:11: note: in instantiation of member function 'hpx::unordered_map<std::__1::basic_string<char>, double, std::__1::hash<std::__1::basic_string<char> >, std::__1::equal_to<std::__1::basic_string<char> > >::register_as' requested here
        m.register_as(map_name);
          ^
/Users/raac/Development/vcpkg/installed/x64-osx/include/hpx/components/containers/unordered/unordered_map.hpp:50:12: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'hpx::unordered_map<std::__1::basic_string<char>, double, std::__1::hash<std::__1::basic_string<char> >, std::__1::equal_to<std::__1::basic_string<char> > >::partitions_vector_type' (aka 'vector<hpx::unordered_map<std::__1::basic_string<char>, double, std::__1::hash<std::__1::basic_string<char> >, std::__1::equal_to<std::__1::basic_string<char> > >::partition_data>') to 'const hpx::server::unordered_map_config_data' for 1st argument
    struct unordered_map_config_data
           ^
/Users/raac/Development/vcpkg/installed/x64-osx/include/hpx/components/containers/unordered/unordered_map.hpp:50:12: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'hpx::unordered_map<std::__1::basic_string<char>, double, std::__1::hash<std::__1::basic_string<char> >, std::__1::equal_to<std::__1::basic_string<char> > >::partitions_vector_type' (aka 'vector<hpx::unordered_map<std::__1::basic_string<char>, double, std::__1::hash<std::__1::basic_string<char> >, std::__1::equal_to<std::__1::basic_string<char> > >::partition_data>') to 'hpx::server::unordered_map_config_data' for 1st argument
    struct unordered_map_config_data
           ^
/Users/raac/Development/vcpkg/installed/x64-osx/include/hpx/components/containers/unordered/unordered_map.hpp:86:9: note: candidate constructor not viable: no known conversion from 'vector<hpx::unordered_map<std::__1::basic_string<char>, double, std::__1::hash<std::__1::basic_string<char> >, std::__1::equal_to<std::__1::basic_string<char> > >::partition_data>' to 'const vector<hpx::server::unordered_map_config_data::partition_data>' for 1st argument
        unordered_map_config_data(std::vector<partition_data> const& partitions)
        ^
/Users/raac/Development/vcpkg/installed/x64-osx/include/hpx/components/containers/unordered/unordered_map.hpp:83:9: note: candidate constructor not viable: requires 0 arguments, but 1 was provided
        unordered_map_config_data()
        ^
/Users/raac/Development/vcpkg/installed/x64-osx/include/hpx/components/containers/unordered/unordered_map.hpp:477:17: error: too many arguments to function call, expected single argument 'symbolic_name', have 2 arguments
                [=](future<id_type> && f) -> future<void> {
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/raac/Development/covid/mobility-cpp/src/connectivity.cpp:24:27: note: in instantiation of member function 'hpx::unordered_map<std::__1::basic_string<char>, double, std::__1::hash<std::__1::basic_string<char> >, std::__1::equal_to<std::__1::basic_string<char> > >::connect_to' requested here
        const auto f1 = m.connect_to(map_name);
                          ^
/Users/raac/Development/vcpkg/installed/x64-osx/include/hpx/runtime/components/client_base.hpp:589:9: note: 'connect_to' declared here
        void connect_to(std::string const& symbolic_name)
        ^
In file included from /Users/raac/Development/covid/mobility-cpp/src/connectivity.cpp:7:
In file included from /Users/raac/Development/vcpkg/installed/x64-osx/include/hpx/include/unordered_map.hpp:10:
/Users/raac/Development/vcpkg/installed/x64-osx/include/hpx/components/containers/unordered/unordered_map.hpp:317:13: error: no matching function for call to 'swap'
            std::swap(partitions_, data.partitions_);
            ^~~~~~~~~
/Users/raac/Development/vcpkg/installed/x64-osx/include/hpx/components/containers/unordered/unordered_map.hpp:329:21: note: in instantiation of member function 'hpx::unordered_map<std::__1::basic_string<char>, double, std::__1::hash<std::__1::basic_string<char> >, std::__1::equal_to<std::__1::basic_string<char> > >::get_data_helper' requested here
                    get_data_helper(id, f.get());
                    ^
/Users/raac/Development/vcpkg/installed/x64-osx/include/hpx/components/containers/unordered/unordered_map.hpp:478:28: note: in instantiation of member function 'hpx::unordered_map<std::__1::basic_string<char>, double, std::__1::hash<std::__1::basic_string<char> >, std::__1::equal_to<std::__1::basic_string<char> > >::connect_to_helper' requested here
                    return connect_to_helper(f.get());
                           ^
/Users/raac/Development/covid/mobility-cpp/src/connectivity.cpp:24:27: note: in instantiation of member function 'hpx::unordered_map<std::__1::basic_string<char>, double, std::__1::hash<std::__1::basic_string<char> >, std::__1::equal_to<std::__1::basic_string<char> > >::connect_to' requested here
        const auto f1 = m.connect_to(map_name);
```

## Steps to Reproduce the Problem

Here's a minimal example that illustrates the issue:

```
#include <hpx/hpx_finalize.hpp>
#include <hpx/hpx_init.hpp>
#include <hpx/include/unordered_map.hpp>
#include <hpx/include/serialization.hpp>

HPX_REGISTER_UNORDERED_MAP(std::string, double);

const char *const map_name = "mobility_visit_map";

int hpx_main() {

    hpx::unordered_map<std::string, double> m;

    if (0 == hpx::get_locality_id()) {
        const auto localities = hpx::find_all_localities();

        m = hpx::unordered_map<std::string, double>(hpx::container_layout(10, localities));
        m.register_as(map_name);
    } else {
        const auto f1 = m.connect_to(map_name);
    }

    return hpx::finalize();
}

int main(int argc, char **argv) {

    std::vector<std::string> const cfg = {
            "hpx.run_hpx_main!=1"
    };

    return hpx::init(argc, argv, cfg);
}
```

## Specifications
  - HPX Version: 1.4.1
  - Platform (compiler, OS): MacOS 10.13.6 (High Sierra). Apple LLVM version 10.0.0 (clang-1000.11.45.5)

--- a/components/containers/partitioned_vector/tests/regressions/CMakeLists.txt
+++ b/components/containers/partitioned_vector/tests/regressions/CMakeLists.txt
@@ -19,7 +19,7 @@ foreach(test ${tests})
     SOURCES ${sources} ${${test}_FLAGS}
     EXCLUDE_FROM_ALL
     HPX_PREFIX ${HPX_BUILD_PREFIX}
-    FOLDER "Tests/Regressions/Components/Containers"
+    FOLDER "Tests/Regressions/Components/Containers/PartitionedVector"
   )
 
   add_hpx_regression_test(

--- a/components/containers/partitioned_vector/tests/unit/CMakeLists.txt
+++ b/components/containers/partitioned_vector/tests/unit/CMakeLists.txt
@@ -38,7 +38,7 @@ foreach(test ${tests})
 
   source_group("Source Files" FILES ${sources})
 
-  set(folder_name "Tests/Unit/Components/Containers")
+  set(folder_name "Tests/Unit/Components/Containers/PartitionedVector")
 
   # add example executable
   add_hpx_executable(

--- a/components/containers/unordered/include/hpx/components/containers/unordered/unordered_map.hpp
+++ b/components/containers/unordered/include/hpx/components/containers/unordered/unordered_map.hpp
@@ -83,6 +83,10 @@ namespace hpx { namespace server
         {}
 
         unordered_map_config_data(std::vector<partition_data> const& partitions)
+          : partitions_(partitions)
+        {}
+
+        unordered_map_config_data(std::vector<partition_data>&& partitions)
           : partitions_(std::move(partitions))
         {}
 
@@ -313,7 +317,12 @@ namespace hpx
         void get_data_helper(id_type id,
             server::unordered_map_config_data data)
         {
-            std::swap(partitions_, data.partitions_);
+            partitions_.clear();
+            partitions_.reserve(data.partitions_.size());
+
+            std::move(data.partitions_.begin(), data.partitions_.end(),
+                std::back_inserter(partitions_));
+
             base_type::reset(std::move(id));
         }
 
@@ -472,8 +481,9 @@ namespace hpx
     public:
         future<void> connect_to(std::string const& symbolic_name)
         {
-            return base_type::connect_to(symbolic_name,
-                [=](future<id_type> && f) -> future<void> {
+            this->base_type::connect_to(symbolic_name);
+            return this->base_type::share().then(
+                [=](shared_future<id_type>&& f) -> hpx::future<void> {
                     return connect_to_helper(f.get());
                 });
         }
@@ -481,13 +491,20 @@ namespace hpx
         // Register this unordered_map with AGAS using the given symbolic name
         future<void> register_as(std::string const& symbolic_name)
         {
-            server::unordered_map_config_data data(partitions_);
+            std::vector<server::unordered_map_config_data::partition_data>
+                partitions;
+            partitions.reserve(partitions_.size());
 
-            base_type::reset(hpx::new_<
-                    typename base_type::server_component_type> >(
-                hpx::find_here(), std::move(data)));
+            std::copy(partitions_.begin(), partitions_.end(),
+                std::back_inserter(partitions));
 
-            return base_type::register_as(symbolic_name);
+            server::unordered_map_config_data data(std::move(partitions));
+            this->base_type::reset(
+                hpx::new_<components::server::distributed_metadata_base<
+                server::unordered_map_config_data>>(
+                    hpx::find_here(), std::move(data)));
+
+            return this->base_type::register_as(symbolic_name);
         }
 
     public:

--- a/components/containers/unordered/tests/regressions/CMakeLists.txt
+++ b/components/containers/unordered/tests/regressions/CMakeLists.txt
@@ -1,5 +1,27 @@
-# Copyright (c) 2019 The STE||AR-Group
+# Copyright (c) 2020 Hartmut Kaiser
 #
 # SPDX-License-Identifier: BSL-1.0
 # Distributed under the Boost Software License, Version 1.0. (See accompanying
 # file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+set(tests unordered_compilation_4685)
+
+set(unordered_compilation_4685_FLAGS COMPONENT_DEPENDENCIES unordered)
+set(unordered_compilation_4685_PARAMETERS THREADS_PER_LOCALITY 2)
+
+foreach(test ${tests})
+  set(sources ${test}.cpp)
+
+  source_group("Source Files" FILES ${sources})
+
+  # add example executable
+  add_hpx_executable(
+    ${test}_test INTERNAL_FLAGS
+    SOURCES ${sources} ${${test}_FLAGS}
+    EXCLUDE_FROM_ALL
+    HPX_PREFIX ${HPX_BUILD_PREFIX}
+    FOLDER "Tests/Regressions/Components/Containers/Unordered"
+  )
+
+  add_hpx_regression_test("components.unordered" ${test} ${${test}_PARAMETERS})
+endforeach()

--- a/components/containers/unordered/tests/regressions/unordered_compilation_4685.cpp
+++ b/components/containers/unordered/tests/regressions/unordered_compilation_4685.cpp
@@ -1,0 +1,46 @@
+//  Copyright (c) 2020 Nick Robison
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#include <hpx/hpx_finalize.hpp>
+#include <hpx/hpx_init.hpp>
+#include <hpx/include/serialization.hpp>
+#include <hpx/include/unordered_map.hpp>
+
+#include <string>
+#include <vector>
+
+HPX_REGISTER_UNORDERED_MAP(std::string, double);
+
+const char* const map_name = "mobility_visit_map";
+
+int hpx_main()
+{
+    hpx::unordered_map<std::string, double> m;
+
+    if (0 == hpx::get_locality_id())
+    {
+        auto localities = hpx::find_all_localities();
+
+        m = hpx::unordered_map<std::string, double>(
+            hpx::container_layout(10, localities));
+        auto f1 = m.register_as(map_name);
+        f1.get();
+    }
+    else
+    {
+        auto f1 = m.connect_to(map_name);
+        f1.get();
+    }
+
+    return hpx::finalize();
+}
+
+int main(int argc, char** argv)
+{
+    std::vector<std::string> const cfg = {"hpx.run_hpx_main!=1"};
+
+    return hpx::init(argc, argv, cfg);
+}

--- a/components/containers/unordered/tests/unit/CMakeLists.txt
+++ b/components/containers/unordered/tests/unit/CMakeLists.txt
@@ -16,7 +16,7 @@ foreach(test ${tests})
 
   source_group("Source Files" FILES ${sources})
 
-  set(folder_name "Tests/Unit/Components/Containers")
+  set(folder_name "Tests/Unit/Components/Containers/Unordered")
 
   # add example executable
   add_hpx_executable(

